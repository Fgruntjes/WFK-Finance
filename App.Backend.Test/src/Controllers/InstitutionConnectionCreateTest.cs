using System.Net;
using System.Net.Http.Json;
using System.Text.Json;
using App.Backend.Dto;
using App.Lib.InstitutionConnection.Exception;
using App.Lib.InstitutionConnection.Service;
using App.Lib.Test;
using Azure.Core;
using Microsoft.AspNetCore.Mvc;
using Moq;

namespace App.Backend.Test.Controllers;

public class InstitutionConnectionCreateTest : IClassFixture<InstitutionConnectionFixture>
{
    private readonly InstitutionConnectionFixture _fixture;

    public InstitutionConnectionCreateTest(InstitutionConnectionFixture fixture)
    {
        _fixture = fixture;
    }

    [Fact]
    public async Task Success()
    {
        // Arrange
        var returnUri = new Uri("http://www.example.com/return");
        _fixture.Services.WithMock<IInstitutionConnectionCreateService>(mock =>
        {
            mock.Setup(e => e.Connect(
                    _fixture.InstitutionEntity.Id,
                    returnUri,
                    It.IsAny<CancellationToken>()))
                .ReturnsAsync(_fixture.InstitutionConnectionEntity);
        });

        // Act
        var response = await _fixture.Client.SendWithAuthAsync(
            new HttpRequestMessage(HttpMethod.Post, "/institutionconnection")
            {
                Content = JsonContent.Create(new InstitutionConnectionCreate
                {
                    InstitutionId = _fixture.InstitutionEntity.Id,
                    ReturnUrl = returnUri
                })
            });
        var body = await response.Content.ReadFromJsonAsync<InstitutionConnection>();

        // Assert
        body.Should().BeEquivalentTo(new InstitutionConnection
        {
            Id = _fixture.InstitutionConnectionEntity.Id,
            ExternalId = _fixture.InstitutionConnectionEntity.ExternalId,
            InstitutionId = _fixture.InstitutionEntity.Id,
            ConnectUrl = _fixture.InstitutionConnectionEntity.ConnectUrl,
            Accounts = new List<InstitutionConnectionAccount>(),
        });
    }

    [Fact]
    public async Task BadRequest_InstitutionNotFoundException()
    {
        // Arrange
        var institutionId = new Guid("8665f9d8-fa7b-497d-b6c5-40fd7e68aefc");
        _fixture.Services.WithMock<IInstitutionConnectionCreateService>(mock =>
        {
            mock.Setup(e => e.Connect(
                   It.IsAny<Guid>(),
                    It.IsAny<Uri>(),
                    It.IsAny<CancellationToken>()))
                .Callback((Guid id, Uri _, CancellationToken _) => throw new InstitutionNotFoundException(id));
        });

        // Act
        var response = await _fixture.Client.SendWithAuthAsync(
            new HttpRequestMessage(HttpMethod.Post, "/institutionconnection")
            {
                Content = JsonContent.Create(new InstitutionConnectionCreate
                {
                    InstitutionId = institutionId,
                    ReturnUrl = new Uri("http://www.example.com/return")
                })
            });

        // Assert
        await response.AssertProblemDetails(HttpStatusCode.BadRequest, new ProblemDetails
        {
            Title = "An error occurred",
            Detail = "Institution {InstitutionId} not found",
            Status = (int)HttpStatusCode.BadRequest,
            Extensions =
            {
                { "InstitutionId", institutionId }
            }
        });
    }
}