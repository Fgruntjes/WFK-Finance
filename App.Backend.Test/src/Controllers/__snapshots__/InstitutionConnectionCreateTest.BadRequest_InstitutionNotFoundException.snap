{
  "errors": [
    {
      "message": "Institution {InstitutionId} not found",
      "locations": [
        {
          "line": 3,
          "column": 9
        }
      ],
      "path": [
        "institutionConnection",
        "create"
      ],
      "extensions": {
        "code": "BAD_REQUEST",
        "timestamp": "2023-12-09T16:14:59.272+01:00",
        "severity": "CRITICAL",
        "metaData": {
          "InstitutionId": "8665f9d8-fa7b-497d-b6c5-40fd7e68aefc"
        }
      }
    }
  ],
  "data": {
    "institutionConnection": {
      "create": null
    }
  }
}
